### PR TITLE
Modify statsclient so that incr works

### PR DESCRIPTION
**Pull request by pchheda-lyft**
Problem: 
Currently in an intelmodule we add two lines 
`from cartography.stats import get_stats_client`
`stat_handler = get_stats_client(__name__)`
When this is done, the base stats client points to None, and hence the stats_client pointed to by the stat_handler is also None
Later when the sync config initializes a new StatsClient, the base stats stats client now points to this object, however, the stats_client pointed to by the stat_handler is still None. 

Point to note: the timeit util function is always called after the sync config  initializes a new StatsClient, and it uses the base stats client, hence those metrics are being published as is.

However, incr doesn't work in the intelmodule.

Existing approach 

Part1 in the screenshot below:
Part 2 in the screenshot below shows why a slightly different approach doesn't work as well
Image from iOS (3)

New approach which works

1. Base static ScopedStatsClient object is created (underlying StatsClient is None, _root is self, prefix is None)
2. When the intelmodule calls stats.get_stats_client, the returned ScopedStatsClient has prefix "intelmodule", _root is object from 1, underlying StatsClient is None
3. When the sync config is applied, and statsd is enabled, the underlying StatsClient for object (_client) from 1 is set to an actual StatsClient object configured to collect metrics
4. When the intelmodule is calling incr on object from 2, the ScopedStatsClient uses _root._client (_root is object from 1 and _root._client was set in step 3), and calls the StatsClient.incr (StatsClient is set! woohoo so the counter metric is published)

I've tested the change in Staging with our internal project, and publishing counter metrics works as expected.

### CODE REVIEW

**Review by achantavy — DISMISSED** [review]
Thanks for working on this!

File: cartography/stats.py [hunk 1]
```diff
@@ -40,10 +28,19 @@ def get_stats_client(self, scope: str) -> 'ScopedStatsClient':
...
+
+        scoped_stats_client = ScopedStatsClient(prefix)
+        scoped_stats_client._parent = self
+        return scoped_stats_client
+
+    def get_base_scoped_stats_client(self) -> 'ScopedStatsClient':
+        base_client = self
+        while base_client._parent:
+            base_client = base_client._parent
+        return base_client
```
Comment (gkleiman) [inline cartography/stats.py:40]:
```suggestion

    def get_root_client() -> 'ScopedStatsClient':
        client = ScopedStatsClient()
        client._root = client

        return client
```

File: cartography/stats.py [hunk 2]
```diff
@@ -40,10 +28,19 @@ def get_stats_client(self, scope: str) -> 'ScopedStatsClient':
...
 
     def is_enabled(self) -> bool:
-        return self._client.is_enabled()
+        return self.get_base_scoped_stats_client()._client is not None
```
Comment (gkleiman) [inline cartography/stats.py:43]:
```suggestion
        return self._root._client is not None
```

File: cartography/stats.py [hunk 3]
```diff
@@ -69,16 +66,16 @@ def timer(self, stat: str, rate: float = 1.0):
...
 
     def set_stats_client(self, stats_client: StatsClient) -> None:
-        self._client.set_client(stats_client)
+        self.get_base_scoped_stats_client()._client = stats_client
 
 
 # Global _scoped_stats_client
 # Will be set when cartography.config.statsd_enabled is True
-_scoped_stats_client: ScopedStatsClient = ScopedStatsClient(_StatsClientWrapper(None))
+_scoped_stats_client: ScopedStatsClient = ScopedStatsClient()
```
Comment (gkleiman) [inline cartography/stats.py:78]:
```suggestion
_scoped_stats_client: ScopedStatsClient = ScopedStatsClient.get_root_client()
```

File: cartography/stats.py [hunk 4]
```diff
@@ -69,16 +66,16 @@ def timer(self, stat: str, rate: float = 1.0):
         if self.is_enabled():
             if self._scope_prefix:
                 stat = f"{self._scope_prefix}.{stat}"
-            return self._client.get_client().timer(stat, rate)
+            return self.get_base_scoped_stats_client()._client.timer(stat, rate)
         return None
 
     def set_stats_client(self, stats_client: StatsClient) -> None:
-        self._client.set_client(stats_client)
+        self.get_base_scoped_stats_client()._client = stats_client
```
Comment (gkleiman) [inline cartography/stats.py:73]:
```suggestion
        self._root._client = stats_client
```

File: cartography/stats.py [hunk 5]
```diff
@@ -69,16 +66,16 @@ def timer(self, stat: str, rate: float = 1.0):
         if self.is_enabled():
             if self._scope_prefix:
                 stat = f"{self._scope_prefix}.{stat}"
-            return self._client.get_client().timer(stat, rate)
+            return self.get_base_scoped_stats_client()._client.timer(stat, rate)
```
Comment (gkleiman) [inline cartography/stats.py:69]:
```suggestion
            return self._root._client.timer(stat, rate)
```

File: cartography/stats.py [hunk 6]
```diff
@@ -56,7 +53,7 @@ def incr(self, stat: str, count: int = 1, rate: float = 1.0) -> None:
         if self.is_enabled():
             if self._scope_prefix:
                 stat = f"{self._scope_prefix}.{stat}"
-            self._client.get_client().incr(stat, count, rate)
+            self.get_base_scoped_stats_client()._client.incr(stat, count, rate)
```
Comment (gkleiman) [inline cartography/stats.py:56]:
```suggestion
            self._root._client.incr(stat, count, rate)
```

**Review by gkleiman — COMMENTED** [review]
What do you think about storing a reference to the root client instead of to the parent?

That way you can avoid having to traverse the whole hierarchy whenever you increment a stat. See the suggestions for how it could look like.

**Review by gkleiman — APPROVED** [review]

**Pull request merged**